python/pyarrow/tests/test_cuda_numba_interop.py

```diff
@@ -75,6 +75,89 @@ def make_random_buffer(size, target='host', dtype='uint8', ctx=None):
     raise ValueError('invalid target value')
 
 
+@pytest.mark.parametrize("c", range(len(context_choice_ids)),
+                         ids=context_choice_ids)
+@pytest.mark.parametrize("dtype", dtypes, ids=dtypes)
+@pytest.mark.parametrize("size", [0, 1, 8, 1000])
+def test_from_object(c, dtype, size):
+    ctx, nb_ctx = context_choices[c]
+    arr, cbuf = make_random_buffer(size, target='device', dtype=dtype, ctx=ctx)
+
+    # Creating device buffer from numba MemoryObject:
+    mem = cbuf.to_numba()
+    cbuf2 = ctx.buffer_from_object(mem)
+    assert cbuf2.size == cbuf.size
+    arr2 = np.frombuffer(cbuf2.copy_to_host(), dtype=dtype)
+    np.testing.assert_equal(arr, arr2)
+
+    # Creating device buffer from a device buffer
+    cbuf2 = ctx.buffer_from_object(cbuf2)
+    assert cbuf2.size == cbuf.size
+    arr2 = np.frombuffer(cbuf2.copy_to_host(), dtype=dtype)
+    np.testing.assert_equal(arr, arr2)
+
+    # Creating device buffer from numba DeviceNDArray:
+    darr = DeviceNDArray((size,), (np.dtype(dtype).itemsize,), np.dtype(dtype),
+                         gpu_data=mem)
+    cbuf2 = ctx.buffer_from_object(darr)
+    assert cbuf2.size == cbuf.size
+    arr2 = np.frombuffer(cbuf2.copy_to_host(), dtype=dtype)
+    np.testing.assert_equal(arr, arr2)
+
+    # Creating device buffer from a slice of numba DeviceNDArray:
+    if size >= 8:
+        for s in [slice(size//4, None, None),
+                  slice(None, -(size//4), None),
+                  slice(size//4, -(size//4), None),
+                  slice(None, None, 2),
+                  slice(size//4, None, 2),
+                  slice(None, -(size//4), 2),
+                  slice(size//4, -(size//4), 2),
+                  slice(None, None, 3),
+                  slice(size//8, None, 3),
+                  slice(None, -(size//8), 3),
+                  slice(size//8, -(size//8), 3)]:
+            s2 = slice(None, None, s.step)
+            cbuf2 = ctx.buffer_from_object(darr[s])
+            arr2 = np.frombuffer(cbuf2.copy_to_host(), dtype=dtype)
+            np.testing.assert_equal(arr[s], arr2[s2])
+
+    # Creating device buffer from a 2-dimensional numba DeviceNDArray:
+    if size >= 8:
+        itemsize = np.dtype(dtype).itemsize
+        darr = DeviceNDArray((size//4, size//(size//4)),
```

Comment:
Instead of creating a new device array (with undefined contents??), just `reshape` the 1-d array.

Reply:
Also, should test a non-contiguous 2d array, and a 2d array with a negative stride.

Reply:
Using reshape. Btw, the new device array had defined content (`gpu_data=mem`).

I'll look into non-contiguous 2d array tests. On can be said beforehand: it will be hard to restore the original array from the returned `CudaBuffer` when the input array is non-contiguous - a copy will be required. Restoring a view would be even impossible.

Reply:
Perhaps [stride_tricks](https://docs.scipy.org/doc/numpy-1.15.0/reference/generated/numpy.lib.stride_tricks.as_strided.html) may help.

Reply:
The problem is that converting a non-contiguous N-D array to a buffer will loose relevant shape and stride information. For array reconstruction from a buffer, exactly the same shape and strides must be used, and take into account that the array and the buffer will have different starting memory addresses. Then on can use stride_tricks indeed.